dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala

```diff
@@ -17,11 +17,35 @@ case class Take[T](self: Exp[_], n: Exp[Int]) extends Exp[TArgClassExample[T]]
 case class X[T](self: Exp[_]) extends Exp[Int]
 case class Y[T](self: Exp[_]) extends Exp[Int]
 case class TArgsZ[T, U](self: Exp[_]) extends Exp[U]
-case class AppCurry[T](self: Exp[_], p1: Exp[T], p2: Exp[T], p3: Exp[T]*) extends Exp[T]
-case class AppManyArgs[T](self: Exp[_], p1: Exp[T]*) extends Exp[T]
+case class AppCurry[T](self: Exp[_], p1: Exp[_], p2: Exp[_], p3: Exp[_]*) extends Exp[T]
+case class AppManyArgs[T](self: Exp[_], p1: Exp[_]*) extends Exp[T]
 
-case class IF[T](cond: Exp[Boolean], e1: Exp[T], e2: Exp[T]) extends Exp[T]
+case class IF[T](cond: Exp[Boolean], e1: Exp[_], e2: Exp[_]) extends Exp[T]
+case class While[T](cond: Exp[Boolean], `do`: Exp[_]) extends Exp[T]
+case class DoWhile[T](cond: Exp[Boolean], `do`: Exp[_]) extends Exp[T]
+case class Return[T](cond: Exp[T]) extends Exp[T]
 case class NewVar[T](e: Exp[T]) extends Exp[T]
+case class ReadVar[T](e: Exp[T]) extends Exp[T]
+case class Assign[T](x: Exp[T], t: Exp[_]) extends Exp[T]
+case class Infix_==[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_Equals[T](t: Exp[T], t1: Exp[_]) extends Exp[T]
+case class Infix_Eq[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_Ne[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_!=[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_##[T](t: Exp[T]) extends Exp[T]
+case class Infix_HashCode[T](t: Exp[T]) extends Exp[T]
+case class Infix_IsInstanceOf[T](t: Exp[_]) extends Exp[T]
+case class Infix_AsInstanceOf[T](t: Exp[_]) extends Exp[T]
+case class Infix_ToString[T](t: Exp[T]) extends Exp[T]
+case class Infix_Notify[T](t: Exp[T]) extends Exp[T]
+case class Infix_NotifyAll[T](t: Exp[T]) extends Exp[T]
+case class Infix_Wait[T](t: Exp[T]) extends Exp[T]
+case class Infix_WaitL[T](t: Exp[T], t1: Exp[_]) extends Exp[T]
+case class Infix_WaitTL[T](t: Exp[T], t1: Exp[_], l: Exp[_]) extends Exp[T]
+case class Infix_Synchronized[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
```

Comment:
I think that the bug is in the definition of the case class only. The 
rest is fine.

On 05/04/15 22:58, Ólafur Páll Geirsson wrote:
>
> In dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala 
> <https://github.com/directembedding/directembedding/pull/28#discussion_r27780443>:
>
> > +case class Infix_==[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
> > +case class Infix_Equals[T](t: Exp[T], t1: Exp[_]) extends Exp[T]
> > +case class Infix_Eq[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
> > +case class Infix_Ne[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
> > +case class Infix_!=[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
> > +case class Infix_##[T](t: Exp[T]) extends Exp[T]
> > +case class Infix_HashCode[T](t: Exp[T]) extends Exp[T]
> > +case class Infix_IsInstanceOf[T](t: Exp[_]) extends Exp[T]
> > +case class Infix_AsInstanceOf[T](t: Exp[_]) extends Exp[T]
> > +case class Infix_ToString[T](t: Exp[T]) extends Exp[T]
> > +case class Infix_Notify[T](t: Exp[T]) extends Exp[T]
> > +case class Infix_NotifyAll[T](t: Exp[T]) extends Exp[T]
> > +case class Infix_Wait[T](t: Exp[T]) extends Exp[T]
> > +case class Infix_WaitL[T](t: Exp[T], t1: Exp[_]) extends Exp[T]
> > +case class Infix_WaitTL[T](t: Exp[T], t1: Exp[_], l: Exp[_]) extends Exp[T]
> > +case class Infix_Synchronized[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
>
> Synchronized will work like this
>
> case  class  Infix_Synchronized[T](self:  Exp[_],t1:Exp[T])extends  Exp[T]
>
> but my gut says it should be more like this
>
> case  class  Infix_Synchronized[T,U](self:  Exp[T],t1:Exp[U])extends  Exp[U]
>
> Should I leave the existential or is this a bug in the transformation? 
> @vjovanov <https://github.com/vjovanov>
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/directembedding/directembedding/pull/28/files#r27780443>.
>